### PR TITLE
[react] Unified type parameter and property types.

### PR DESCRIPTION
**Pull request by DovydasNavickas**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [x] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

`type` parameter and property types were updated in several places, but not all of them.
I've limited the ones that were left from `string` to `keyof ReactHTML` and incremented TypeScript version to 2.4.

### CODE REVIEW

**Review by vsiao — CHANGES_REQUESTED** [review]
This is incorrect—SVG elements are DOM elements too.

**Comment by DovydasNavickas** [comment]
Yep. I've updated `ReactElement` types to accept SVG also and get type-gates for specific ones.

In case of `createElement` there is an overload for SVG with `keyof ReactSVG` type:
https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/react/index.d.ts#L144

So, now [ReactElement][0] accepts everything in its `type`:
```ts
interface ReactElement<P> {
    type: keyof ReactHTML | keyof ReactSVG | ComponentClass<P> | SFC<P>;
    // <...>
}
```

[SFCElement][1] limits to only accept `SFC<P>`:
```ts
interface SFCElement<P> extends ReactElement<P> {
    type: SFC<P>;
}
```
[ComponentElement][2] limits to only accept `ComponentClass<P>`:
```ts
interface ComponentElement<P, T extends Component<P, ComponentState>> extends ReactElement<P> {
    type: ComponentClass<P>;
    // <...>
}
```

[DOMElement][3] can be either HTML or SVG element, as you said, which is now true:
```ts
interface DOMElement<P extends DOMAttributes<T>, T extends Element> extends ReactElement<P> {
    type: keyof ReactHTML | keyof ReactSVG;
    ref: Ref<T>;
}
```

And then for [ReactHTMLElement][4] and [ReactSVGElement][5] `type` is scoped to be specific as so:
```ts
interface ReactHTMLElement<T extends HTMLElement> extends DOMElement<HTMLAttributes<T>, T> {
    type: keyof ReactHTML;
}

interface ReactSVGElement extends DOMElement<SVGAttributes<SVGElement>, SVGElement> {
    type: keyof ReactSVG;
}
```

`createElement` functions now look good also:
https://github.com/DovydasNavickas/DefinitelyTyped/blob/354aa9339ccb709e83695850aa7d0e2fa79ecc86/types/react/index.d.ts#L142-L165

When given type is `keyof ReactHTML`, props have to extend `HTMLAttributes` and element has to extend `HTMLElement`, which then return `ReactHTMLElement`:
```ts
function createElement<P extends HTMLAttributes<T>, T extends HTMLElement>(
        type: keyof ReactHTML,
        props?: ClassAttributes<T> & P,
        ...children: ReactNode[]): ReactHTMLElement<T>;
```

Corresponding overloads are for `SVG`, `SFC`, `CElement`. If one gives `ComponentClass` as a type, `ReactElement` is returned.

I'm just not sure about the factories here:
https://github.com/DovydasNavickas/DefinitelyTyped/blob/354aa9339ccb709e83695850aa7d0e2fa79ecc86/types/react/index.d.ts#L106-L113

I think `HTMLElement` and `SVGElement` types already scope them enough. Any objections? Did I miss something?

[0]: https://github.com/DovydasNavickas/DefinitelyTyped/blob/354aa9339ccb709e83695850aa7d0e2fa79ecc86/types/react/index.d.ts#L52
[1]: https://github.com/DovydasNavickas/DefinitelyTyped/blob/354aa9339ccb709e83695850aa7d0e2fa79ecc86/types/react/index.d.ts#L58
[2]: https://github.com/DovydasNavickas/DefinitelyTyped/blob/354aa9339ccb709e83695850aa7d0e2fa79ecc86/types/react/index.d.ts#L63
[3]: https://github.com/DovydasNavickas/DefinitelyTyped/blob/354aa9339ccb709e83695850aa7d0e2fa79ecc86/types/react/index.d.ts#L70
[4]: https://github.com/DovydasNavickas/DefinitelyTyped/blob/354aa9339ccb709e83695850aa7d0e2fa79ecc86/types/react/index.d.ts#L75
[5]: https://github.com/DovydasNavickas/DefinitelyTyped/blob/354aa9339ccb709e83695850aa7d0e2fa79ecc86/types/react/index.d.ts#L79

**Comment by DovydasNavickas** [comment]
Ooook. That was a tough ride.

I've added `tslint.json` for `react` and fixed all errors (with a few exceptions that needed `tslint:disable-next-line`, such as empty interfaces).
As a side error fixing, I've added `tslint.json` for `JSnoX` package and fixed the errors.

Also, added required `React.cloneElement` overloads for `ReactHTMLElement` and `ReactSVGElement` and appropriate tests.

There are quite a few changes because of `tslint` error fixing. In essence, the changes are these:
https://github.com/DovydasNavickas/DefinitelyTyped/blob/a1b1c00e68e9648f08544ad8c0609737c7dd4c09/types/react/index.d.ts#L72-L83
And here:
https://github.com/DovydasNavickas/DefinitelyTyped/blob/a1b1c00e68e9648f08544ad8c0609737c7dd4c09/types/react/index.d.ts#L141-L192

@sandersn, I think your changes in https://github.com/DefinitelyTyped/DefinitelyTyped/pull/17591 are not correct.
With the changes I've made now there is a constraint on `type` string and it cannot be just any string.
It has to be `ReactHTML` or `ReactSVG` value or it has to be a component class otherwise. 

This PR also resolves https://github.com/DefinitelyTyped/DefinitelyTyped/pull/17507 that you addressed.

Please review and comment if you see anything odd.

**Review by MartynasZilinskas — APPROVED** [review]

**Review by DeividasBakanas — APPROVED** [review]

**Comment by DovydasNavickas** [comment]
I've also added `string` fallback for `ReactElement` and `DOMElement`. Now d.ts supports everything:
`type: ReactHTML` => `ReactHTMLElement`
`type: ReactSVG` => `ReactSVGElement`
`type: string` => `DOMElement`

I really hope this ends this ode to React 😄

File: types/react/index.d.ts
Comment (sandersn) [inline types/react/index.d.ts]: this addition doesn't do anything since `string` is a supertype of `keyof ReactHTML` and `keyof ReactSVG`.

File: types/react/index.d.ts
Comment (sandersn) [inline types/react/index.d.ts]: same here

File: types/react/index.d.ts
Comment (sandersn) [inline types/react/index.d.ts]: same here

File: types/react/index.d.ts
Comment (sandersn) [inline types/react/index.d.ts]: Does this mean `createFactory` no longer supports SVG and custom types?

File: types/react/index.d.ts
Comment (sandersn) [inline types/react/index.d.ts]: the choice of SVGAttributes seems wrong for the fallback overload.

File: types/react/index.d.ts
Comment (sandersn) [inline types/react/index.d.ts]: I don't know enough about React to understand this change. Anyone else?

File: types/react/test/index.ts
Comment (sandersn) [inline types/react/test/index.ts]: does the fallback case still work? Why remove this test?

**Review by sandersn — CHANGES_REQUESTED** [review]

File: types/react/index.d.ts
Comment (DovydasNavickas) [inline types/react/index.d.ts]: Leaving `P` here is more correct.

File: types/react/index.d.ts
Comment (DovydasNavickas) [inline types/react/index.d.ts]: Fixed: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/17553/commits/929b4a206daefb5463be587e74ea5fa225c40e90#diff-96b72df8b13a8a590e4f160cbc51f40cR55

File: types/react/index.d.ts
Comment (DovydasNavickas) [inline types/react/index.d.ts]: Fixed: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/17553/commits/929b4a206daefb5463be587e74ea5fa225c40e90#diff-96b72df8b13a8a590e4f160cbc51f40cR74

File: types/react/index.d.ts
Comment (DovydasNavickas) [inline types/react/index.d.ts]: Added missing overloads for `createFactory`:
https://github.com/DefinitelyTyped/DefinitelyTyped/pull/17553/commits/929b4a206daefb5463be587e74ea5fa225c40e90#diff-96b72df8b13a8a590e4f160cbc51f40cR136

**Review by sandersn — APPROVED** [review]

File: types/react/index.d.ts
Comment (DovydasNavickas) [inline types/react/index.d.ts]: Fixed: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/17553/commits/929b4a206daefb5463be587e74ea5fa225c40e90#diff-96b72df8b13a8a590e4f160cbc51f40cR160

File: types/react/index.d.ts
Comment (DovydasNavickas) [inline types/react/index.d.ts]: Fixed:
https://github.com/DefinitelyTyped/DefinitelyTyped/pull/17553/commits/929b4a206daefb5463be587e74ea5fa225c40e90#diff-96b72df8b13a8a590e4f160cbc51f40cR161

File: types/react/index.d.ts
Comment (DovydasNavickas) [inline types/react/index.d.ts]: Fixed: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/17553/commits/929b4a206daefb5463be587e74ea5fa225c40e90#diff-96b72df8b13a8a590e4f160cbc51f40cR163

File: types/react/test/index.ts
Comment (DovydasNavickas) [inline types/react/test/index.ts]: Readed the test with a correct definition: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/17553/commits/929b4a206daefb5463be587e74ea5fa225c40e90#diff-96b72df8b13a8a590e4f160cbc51f40cR163

**Comment by DovydasNavickas** [comment]
@sandersn, build succeeded, so it's good to merge? 😄

**Comment by sandersn** [comment]
I hoped that one of the React maintainers would have a chance to take a look too. If not, I'll merge tomorrow.

**Comment by DovydasNavickas** [comment]
@sandersn so?

**Comment by sandersn** [comment]
How urgent is this? When I'm working on DefinitelyTyped I like to wait 3 days before merging something to make sure maintainers have a chance to look at it, since they are all volunteers. And react is much more important than most libraries, so unless it's outright broken, it's better to be cautious.

**Comment by DovydasNavickas** [comment]
Well, latest changes before this PR broke our builds, and it's more of "finishing the job" than "introducing new changes".
Most of the changes came from `tslint.json` introduction.

I have disabled the tests for our build at the moment, because it's hard not to use the latest version of `React` d.ts when all of the other @types also use the latest ones.

The problem could only be encountered if one would be using `React` without `JSX` or augmenting the result of JSX, e.g. like so:
```tsx
import * as React from "react";

const tagsList = [
    <meta charSet="utf-8" />,
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
];

function GetTags(keyPrefix: string): JSX.Element[] {
    return tagsList.map((tag, index) => {
        const props = { ...tag.props, ...{ key: `${keyPrefix}-${index}` } };

        // --- This is where it failed on tag.type not fitting the createElement signature
        return React.createElement(tag.type, props);
    }).filter(x => x != null) as JSX.Element[];
}
```

Therefore, I guess we're one of a few who use this kind of stuff and can wait for more eyes on PR :)

**Comment by DovydasNavickas** [comment]
Hey, @sandersn :) Judging from the silence, I guess there will be no more feedback. Can you remove the tag `The Travis CI build failed` and merge? :)

**Comment by DovydasNavickas** [comment]
Thanks! 🎉

**Pull request merged**
